File: packages/domain/src/stats.ts
import { GameCode, type DrawRecord, type StatsResponseContract } from "@lottery/contracts";
import { getGameDefinition, listInclusiveRange } from "./games";

function frequency(draws: DrawRecord[], key: "primaryNumbers" | "secondaryNumbers"): Record<number, number> {
  const counter = new Map<number, number>();
  for (const draw of draws) {
    for (const number of draw[key]) {
      counter.set(number, (counter.get(number) ?? 0) + 1);
    }
  }
  return Object.fromEntries([...counter.entries()].sort((left, right) => left[0] - right[0]));
}

function missingPeriods(
  draws: DrawRecord[],
  key: "primaryNumbers" | "secondaryNumbers",
  range: [number, number],
): Record<number, number> {
  const missing = new Map<number, number>();
  for (const number of listInclusiveRange(range)) {
    missing.set(number, draws.length);
  }
  draws.forEach((draw, index) => {
    for (const number of draw[key]) {
      if ((missing.get(number) ?? draws.length) === draws.length) {
        missing.set(number, index);
      }
    }
  });
  return Object.fromEntries([...missing.entries()].sort((left, right) => left[0] - right[0]));
}

export function buildStats(gameCode: GameCode, draws: DrawRecord[], dataVersion: string): StatsResponseContract {
  const definition = getGameDefinition(gameCode);
  return {
    gameCode,
    drawCount: draws.length,
    latestIssue: draws[0]?.issue ?? null,
    primaryFrequency: frequency(draws, "primaryNumbers"),
    secondaryFrequency: frequency(draws, "secondaryNumbers"),
    primaryMissingPeriods: missingPeriods(draws, "primaryNumbers", definition.primaryRange),
    secondaryMissingPeriods: missingPeriods(draws, "secondaryNumbers", definition.secondaryRange),
    dataVersion,
  };
}

File: server/app/services/stats.py
from collections import Counter

from app.db.models import Draw
from app.domain.games import GameCode, get_game_definition


def _frequency(draws: list[Draw], attr: str) -> dict[int, int]:
    counter: Counter[int] = Counter()
    for draw in draws:
        counter.update(getattr(draw, attr))
    return dict(sorted(counter.items()))


def _missing_periods(draws: list[Draw], attr: str, number_range: range) -> dict[int, int]:
    missing = {number: len(draws) for number in number_range}
    for index, draw in enumerate(draws):
        for number in getattr(draw, attr):
            if missing[number] == len(draws):
                missing[number] = index
    return dict(sorted(missing.items()))


def build_stats(game_code: GameCode, draws: list[Draw], data_version: str) -> dict:
    definition = get_game_definition(game_code)
    return {
        "game_code": game_code,
        "draw_count": len(draws),
        "latest_issue": draws[0].issue if draws else None,
        "primary_frequency": _frequency(draws, "primary_numbers"),
        "secondary_frequency": _frequency(draws, "secondary_numbers"),
        "primary_missing_periods": _missing_periods(
            draws,
            "primary_numbers",
            definition.primary_range,
        ),
        "secondary_missing_periods": _missing_periods(
            draws,
            "secondary_numbers",
            definition.secondary_range,
        ),
        "data_version": data_version,
    }

